src/PoolManager.sol

```diff
@@ -203,34 +204,50 @@ contract PoolManager is IPoolManager, ProtocolFees, NoDelegateCall, ERC6909Claim
         PoolId id = key.toId();
         _checkPoolInitialized(id);
 
-        key.hooks.beforeSwap(key, params, hookData);
+        int128 hookDeltaInSpecified;
+        (params.amountSpecified, hookDeltaInSpecified) = key.hooks.beforeSwap(key, params, hookData);
 
-        uint256 feeForProtocol;
-        uint24 swapFee;
-        Pool.SwapState memory state;
-        (delta, feeForProtocol, swapFee, state) = pools[id].swap(
+        // execute swap, account protocol fees, and emit swap event
+        delta = _swap(
+            id,
             Pool.SwapParams({
                 tickSpacing: key.tickSpacing,
                 zeroForOne: params.zeroForOne,
                 amountSpecified: params.amountSpecified,
                 sqrtPriceLimitX96: params.sqrtPriceLimitX96
-            })
+            }),
+            params.zeroForOne ? key.currency0 : key.currency1
         );
 
-        _accountPoolBalanceDelta(key, delta);
+        (int128 hookDeltaInUnspecified) = key.hooks.afterSwap(key, params, delta, hookData);
+
+        // calculates if currency0 or currency1 is the specified token
+        BalanceDelta hookDelta = ((params.amountSpecified < 0) == params.zeroForOne)
```

Comment:
Using `params.amountSpecified < 0` to determine if the swap was exactInput would break in NoOp cases (`hookDeltaInSpecified == params.amountSpecified`)